eval/compile_effect.go

```diff
@@ -358,23 +328,35 @@ func allTrue(vs []interface{}) bool {
 	return true
 }
 
-func (cp *compiler) assignment(n *parse.Assignment) effectOpBody {
+func (cp *compiler) assignment(n *parse.Assignment, temporary bool) effectOpBody {
```

Comment:
Overall, what you are doing here is not quite enough; because Elvish checks for variable existence statically, code like `x=foo echo $x; echo $x` should be a **compilation error**. This means that deleting the variable from the runtime Ns is not sufficient, it also has to be deleted from the compiler's scope after compiling the first form `x=foo echo $x`, so that the second form `echo $x` does not compile at all.

Essentially what you have to do is to revert the side effect `registerVariableSet` call [here](https://github.com/elves/elvish/blob/71d050b31c1578b763cf9d6cdae50ca6fcc2956a/eval/compile_lvalue.go#L81), for variables that did not exist before the temporary assignment. Doing so is non-trivial (which is why I did not implement it in the first place)

Reply:
Ah, I see what you mean. I felt I'm digging a rabbit hole when I started. Should we make an annotation on the temporary variable and let the compiler know when it tries to get the variable from up scope? SInce reverting `registerVariableSet` in compiler will cause runtime failure.

Reply:
I feel that you took "reverting `registerVariableSet`" a bit too literally; I don't mean removing that call, but something a bit more complex:

1. When doing a temporary assignment, you need to record which new variables have been created in the scope as a side effect of that `registerVariableSet` call [here](https://github.com/elves/elvish/blob/71d050b31c1578b763cf9d6cdae50ca6fcc2956a/eval/compiler.go#L138)

2. After the entire form has been compiled, remove those new variables from the scope.

Step 2 is actually a parallel process to how you delete the name from the `Ns` when reverting the temporary assignment. The compilation process needs to mirror what happens in the runtime.

I don't think it's undoable, but because the reversion in the compilation process needs to happen in the `form` method, and `registerVariableSet` is buried several call stacks deep under, letting the `form` method know which variables are newly set and should be reverted involves either a) passing that information all the way up (meh), or b) recording it somewhere in the compiler (meh and worse). Both are kind of painful, but approach a is the lesser evil here.